test/e2e/annotations/canary.go

```diff
@@ -832,4 +875,211 @@ var _ = framework.DescribeAnnotation("canary", func() {
+				"nginx.ingress.kubernetes.io/session-cookie-name": affinityCookieName,
+			}
+
+			ing := framework.NewSingleIngress(host, "/", host,
+				f.Namespace, framework.EchoService, 80, annotations)
+			f.EnsureIngress(ing)
+
+			f.WaitForNginxServer(host,
+				func(server string) bool {
+					return strings.Contains(server, "server_name foo")
+				})
+
+			// Canary weight is 50% to ensure requests are going there.
+			// Explicitly set affinity-canary-behavior annotation to "legacy".
+			canaryAnnotations := map[string]string{
+				"nginx.ingress.kubernetes.io/canary":                   "true",
+				"nginx.ingress.kubernetes.io/canary-by-header":         "ForceCanary",
+				"nginx.ingress.kubernetes.io/canary-by-header-value":   "yes",
+				"nginx.ingress.kubernetes.io/canary-weight":            "50",
+				"nginx.ingress.kubernetes.io/affinity-canary-behavior": "legacy",
+			}
+
+			canaryIng := framework.NewSingleIngress(canaryIngName, "/", host,
+				f.Namespace, canaryService, 80, canaryAnnotations)
+			f.EnsureIngress(canaryIng)
+
+			// This request will produce affinity cookie coming from the canary
+			// backend.
+			forcedRequestToCanary := f.HTTPTestClient().
+				GET("/").
+				WithHeader("Host", host).
+				WithHeader("ForceCanary", "yes").
+				Expect().
+				Status(http.StatusOK)
+
+			// Make sure we got response from canary.
+			forcedRequestToCanary.
+				Body().Contains(canaryService)
+
+			// Legacy behavior results in affinity cookie not being set in
+			// response.
+			for _, c := range forcedRequestToCanary.Cookies().Iter() {
+				if c.String().Raw() == affinityCookieName {
+					ginkgo.GinkgoT().Error("Affinity cookie is present in response, but was not expected.")
+				}
+			}
+
+			TestEvenMainlineCanaryDistribution(f, host)
+		})
+	})
+
 })
+
+// This method assumes canary weight being configured at 50%.
+func TestEvenMainlineCanaryDistribution(f *framework.Framework, host string) {
+	re := regexp.MustCompile(fmt.Sprintf(`%s.*`, framework.EchoService))
+	replicaRequestCount := map[string]int{}
+
+	for i := 0; i < 200; i++ {
+		body := f.HTTPTestClient().
+			GET("/").
+			WithHeader("Host", host).
+			Expect().
+			Status(http.StatusOK).Body().Raw()
+
+		replica := re.FindString(body)
+		assert.NotEmpty(ginkgo.GinkgoT(), replica)
+
+		if _, ok := replicaRequestCount[replica]; !ok {
+			replicaRequestCount[replica] = 1
+		} else {
+			replicaRequestCount[replica]++
+		}
+	}
+
+	keys := reflect.ValueOf(replicaRequestCount).MapKeys()
+
+	assert.Equal(ginkgo.GinkgoT(), 2, len(keys))
+
+	// The implmentation of choice by weight doesn't guarantee exact
+	// number of requests, so verify if request imbalance is within an
+	// acceptable range.
+	assert.LessOrEqual(ginkgo.GinkgoT(), math.Abs(float64(replicaRequestCount[keys[0].String()]-replicaRequestCount[keys[1].String()]))/math.Max(float64(replicaRequestCount[keys[0].String()]), float64(replicaRequestCount[keys[1].String()])), 0.2)
```

Comment:
being honest, I'm  afraid this generates some flakes, but let's watch

Reply:
Yeah, I ended up making a margin of error 20% for the "even" distribution. Not happy, but making proper round-robin for this case would be a much bigger change.

Reply:
Hello @rikatz and @wasker , I can see on the latest e2e run for K8s 1.21, [this test is failing] (https://github.com/kubernetes/ingress-nginx/runs/5422732839?check_suite_focus=true) with error `"0.23008849557522124" is not less than or equal to "0.2"`
Also, locally I had this failling  <->  not-failling inconsistently. 
If even distribution is not exactly even, should we keep maintaining this test?